eclipse-scout-core/src/dataobject/serialize/DoSerializer.ts

```diff
@@ -0,0 +1,139 @@
+    this.cycleDetector.push(value);
+    try {
+      const serializer = dataObjects.serializers.find(s => s.canSerialize(value, valueMetaData));
+      if (serializer) {
+        // use custom serializer
+        return serializer.serialize(value, valueMetaData, this);
+      }
+      if (objects.isNullOrUndefined(value)) {
+        return value;
+      }
+      if (objects.isObject(value)) {
+        // nested object
+        return this._serializeObject(value, valueMetaData);
+      }
+      return value;
+    } finally {
+      this.cycleDetector.pop(value);
+    }
+  }
+
+  protected _serializeObject(value: any, metaData?: DoValueMetaData): any {
+    const detectedObjectType = this._detectObjectType(value);
+    let objectType = metaData?.type;
+    if (objectType) {
+      doValueMetaData.assertTypesCompatible(detectedObjectType, objectType);
+    } else {
+      objectType = detectedObjectType;
+    }
+
+    const target = Object.assign({}, value); // shallow copy to keep original object intact
+    this._setJsonTypeTo(target, objectType);
+    delete target.objectType; // Scout JS internal property
+
+    const proto = objectType ? Object.getPrototypeOf(objectType).prototype : null;
+    Object.keys(target)
+      .filter(key => key !== '_type')
+      .forEach(key => {
+        target[key] = this._convertFieldValue(proto, key, target[key]);
+      });
+    return target;
+  }
+
+  protected _detectObjectType(obj: any): Constructor {
+    if (obj instanceof BaseDoEntity) {
+      return obj.constructor as Constructor;
+    }
+    const constructor = doValueMetaData.resolveToConstructor(obj.objectType) as Constructor;
+    if (constructor) {
+      return constructor;
+    }
+    if (typeof obj._type === 'string') {
+      return DoRegistry.get().toConstructor(obj._type);
+    }
+    return null;
+  }
+
+  protected _convertFieldValue(proto: object, fieldName: string, value: any): any {
+    this.cycleDetector.pushAttributeName(fieldName);
+    const fieldMetaData = doValueMetaData.getFieldMetaData(proto, fieldName);
+    const result = this.serialize(value, fieldMetaData);
+    this.cycleDetector.popAttributeName(fieldName);
+    return result;
+  }
+
+  protected _setJsonTypeTo(obj: any, constructor: Constructor) {
+    if (obj._type) {
+      return; // already present
+    }
+    const objectFactory = ObjectFactory.get();
+    const objectType = objectFactory.getObjectType(constructor) || objectFactory.getObjectType(obj.objectType);
+    if (objectType) {
+      const jsonType = DoRegistry.get().toJsonType(objectType);
+      if (jsonType) {
+        obj._type = jsonType;
+      }
+    }
+  }
+}
+
+export class DoSerializerCycleDetector {
+  protected _stack = [];
+  protected _attributeNames: string[] = [];
+
+  push(value: any) {
+    if (!this._canHaveChildren(value)) {
+      return;
+    }
+    if (this.contains(value)) {
+      const path = '[' + strings.join(',', ...this._attributeNames) + ']';
```

Comment:
`this._attributeNames.join(',')` ?

Reply:
Yes, fine with me. But is the behavior of the native join function really the same as the one from Scout?
Especially because the scout version uses strings.asString which has some extra logic?